component/src/lib.rs

```diff
@@ -33,81 +31,82 @@ pub trait Manageable {
     async fn spawn(mut self) -> Result<(), anyhow::Error> {
-        let (tx_panic, tr_panic) = channel(10);
-        let (tx_shutdown, tr_shutdown) = oneshotChannel();
+        let (tx_irrecoverable, tr_irrecoverable) = channel(10);
+        let (tx_cancellation, tr_cancellation) = oneshotChannel();
 
         // this assumes the start_fn is populated
-        let wrapped_handle = self.manageable.start(tx_panic, tr_shutdown).await;
+        let wrapped_handle = self.manageable.start(tx_irrecoverable, tr_cancellation).await;
 
-        self.panic_signal = tr_panic;
-        self.shutdown_signal = tx_shutdown;
+        self.irrecoverable_signal = tr_irrecoverable;
+        self.cancellation_signal = tx_cancellation;
         self.join_handle = Some(wrapped_handle);
 
-        // we're ourselves a component
         self.run().await
     }
 
     // Run supervision of the child task
     async fn run(mut self) -> Result<(), anyhow::Error> {
-        // TODO : how does this conflict with join ? Do we still want join?
-        // I don't think we need to join, because we never expect these tasks to complete
-        // exception is on application shutdown, which theoretically should never happen.
-        //
         // select statement that listens for the following cases:
         //
-        // panic signal incoming => log, terminate and restart
+        // Irrecoverable signal incoming => log, terminate and restart
         // completion of the task => we're done! return
 
         if let Some(mut handle) = self.join_handle {
             loop {
                 tokio::select! {
-                    Some(message) = self.panic_signal.recv() => {
+                    Some(message) = self.irrecoverable_signal.recv() => {
 
-                        println!("panic message received: {:?}", message.to_string());
                         self.manageable.handle_irrecoverable(message).await;
                         self.terminate().await?;
 
-                        // restart, continuous passing style
-                        let next_component = Component::new(self.manageable.clone());
-                        self = next_component;
-                        self.spawn();
+                        // restart
+                        let (panic_sender, panic_receiver) = channel(10);
+                        let (cancel_sender, cancel_receiver) = oneshotChannel();
+
+                        let wrapped_handle: tokio::task::JoinHandle<()> =  self.manageable.start(panic_sender, cancel_receiver).await;
+
+                        self.irrecoverable_signal = panic_receiver;
+                        self.cancellation_signal = cancel_sender;
+                        self.join_handle = Some(wrapped_handle);
+
+                        // // restart, continuous passing style
+                        // let next_component = Supervisor::new(self.manageable.clone());
+                        // self = next_component;
+                        // self.spawn().await;
```

Comment:
This continuous passing does not work because it violates the ownership, but I could be missing something.

Reply:
At this stage, I don't think we need it: the task is de facto restarted when calling `manageable.start` above.